bundles/org.openhab.binding.viessmann/src/main/java/org/openhab/binding/viessmann/internal/handler/ViessmannBridgeHandler.java

```diff
@@ -0,0 +1,572 @@
+        } else {
+            apiCalls = 0;
+        }
+        newInstallationId = "";
+        newGatewaySerial = "";
+
+        api = new ViessmannApi(this.config.apiKey, httpClient, this.config.user, this.config.password,
+                this.config.installationId, this.config.gatewaySerial, callbackUrl);
+        api.createOAuthClientService(this);
+
+        if (api.doAuthorize()) {
+            if (this.config.installationId.isEmpty() || this.config.gatewaySerial.isEmpty()) {
+                api.setInstallationAndGatewayId(this);
+                setConfigInstallationGatewayId();
+            }
+
+            if (!config.disablePolling && errorChannelsLinked()) {
+                startViessmannErrorsPolling(config.pollingIntervalErrors);
+            }
+
+            migrateChannelIds();
+
+            getAllDevices();
+            if (!devicesList.isEmpty()) {
+                updateBridgeStatus(ThingStatus.ONLINE);
+                startViessmannBridgePolling(getPollingInterval(), 1);
+            }
+        }
+    }
+
+    private void migrateChannelIds() {
+        List<Channel> oldChannels = thing.getChannels();
+        List<Channel> newChannels = new ArrayList<>(oldChannels.size());
+
+        Map<ChannelUID, ChannelUID> renameMap = new LinkedHashMap<>();
+
+        for (Channel channel : oldChannels) {
+            String oldId = channel.getUID().getId();
+            String newId = ViessmannUtil.camelToHyphen(oldId);
+            boolean updateChannelType = false;
+            String channelLabel = channel.getLabel();
+            String channelType = ViessmannUtil
+                    .camelToHyphen(Objects.requireNonNull(channel.getChannelTypeUID()).toString());
+            channelType = channelType.replace(BINDING_ID + ":", "");
+
+            if (channelType.contains("type-")) {
+                channelType = channelType.replace("type-", "");
+                updateChannelType = true;
+            }
+
+            if (!newId.equals(oldId) || updateChannelType) {
+                logger.info("Migrating channel '{}' -> '{}'", oldId, newId);
+
+                ChannelUID oldUid = channel.getUID();
+                ChannelUID newUid = new ChannelUID(thing.getUID(), newId);
+
+                ChannelTypeUID channelTypeUID = new ChannelTypeUID(BINDING_ID, channelType);
+                if (channelLabel != null) {
+                    Channel newChannel = ChannelBuilder.create(newUid, channel.getAcceptedItemType())
+                            .withLabel(channelLabel).withType(channelTypeUID).withProperties(channel.getProperties())
+                            .build();
+
+                    newChannels.add(newChannel);
+                    renameMap.put(oldUid, newUid);
+                }
+            }
+        }
+
+        if (renameMap.isEmpty()) {
+            return;
+        }
+
+        updateThing(editThing().withChannels(newChannels).build());
+
+        if (linkRegistry != null) {
```

Comment:
i don't think the linkRegistry can be null. It is marked nonnull and injected from the framework.